Makefile edit to build on macos

bjornbm opened the pull request:
NOT READY FOR MERGE. Disclaimer: I'm not at all a Makefile or C toolchain expert.

These are changes I had to make to be able to puild pt_lua and run tests on macos.

The second commit (`-export_dynamic`) you should not taken as-is. It is to illustrate what I needed to do for Apple ld, but I guess it will break GNU ld which uses `-E` or `--export_dynamic`. Can the right flag be detected/selected conditionally without dragging in all of autotools?

bjornbm commented:
I realize now that maybe `LDFLAGS` should not be set with the `-L…` so that it can be used for `LDFLAGS=-lasan` per the README.

If so, even after reverting `PTLUA_LDFLAGS`, I'm 95% sure that `-L$(LUA_LIBDIR)` needs to be added to the `.so` target to work on my mac.

hugomg commented:
Indeed, "-L" is part of the flags, and should go inside the definition of LDFLAGS

Before we edit the makefile, do you know if there is a way to get what you want by configuring flags when you call make? 
For example: `make PTLUA_LDFLAGS='-Wl,-export_dynamic'` or something along those lines?

In the long term I would love to come up with a simpler way to compile Pallene itself as well as Pallene programs. But so far I haven't cracked that puzzle.

bjornbm commented:
You are of course correct that I can override `PTLUA_LDFLAGS` when running make, exactly as you showed. This seems to be the pragmatic solution since the Xcode linker is the odd one out (versus GNU ld, gold, and LLVM lld…).

See the related addition in the README.

Regarding `-L$(LUA_LIBDIR)`, while in principle I agree it should go in `LDFLAGS`, I propose to keep it separate to reduce the incidental “noise” when a user is building on macos or needs to set the `-lasan` flag. I simply inlined the `-L` in the `$(CC)` commands, but if you don't like that maybe a dedicated variable is better (`LDFLAGS_LUA`??)

I also documented how to run the tests (although maybe should mention that busted is required?).

hugomg commented:
I think the idea is first we figure out how to do it with just flags, and possibly document it. 

Then in the second step we figure out a cleaner user interface. For example, Lua's makefile has "make linux" and "make macosx"

My main worry is that don't want to mix linux & macosx in the same command line. They should each get different ldflags etc.

bjornbm commented:
I don't disagree. But it is not clear to me if the current diffs look like a good-enough approach for now: https://github.com/pallene-lang/pallene-tracer/pull/27/files

If we settle, I can do a new merge request with cleaner commits (without the back-and-forth) if you prefer.

singul4ri7y commented on Makefile:
```diff
@@ -27,7 +27,7 @@ LIBFLAG  = -fPIC -shared
 # The -Wl,-E tells the linker to not throw away unused Lua API symbols.
 # We need them for Lua modules that are dynamically linked via require
-PTLUA_LDFLAGS = -L$(LUA_LIBDIR) -Wl,-E
+PTLUA_LDFLAGS = -Wl,--export_dynamic
```
I have a suggestions here:

We should pass Linker flags to `make` with `LDFLAGS`, not via `PTLUA_LDFLAGS` directly. So, to make it all work, we can define the `LDFLAGS` in the Makefile instead of keeping it empty and retain `-L$(LUA_LIBDIR)` in `PTLUA_LDFLAGS`: 
 
 ```Makefile
PTLUA_LDFLAGS = -L$(LUA_LIBDIR)
LDFLAGS = -Wl,-E
 ```

Hence, while building for MacOS, we can do the same thing as @hugomg suggested but with `LDFLAGS`,
```
make LDFLAGS='-Wl,--export_dynamic'
```

singul4ri7y commented on Makefile:
```diff
@@ -65,10 +65,10 @@ clean:
 
 %.so: %.c
-	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) $(LIBFLAG) $< -o $@
+	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) -L$(LUA_LIBDIR) $(LIBFLAG) $< -o $@ $(PTLUA_LDLIBS)
```
We should move `-L$(LUA_LIBDIR)` to `PTLUA_LDFLAGS` as it is compatible with GNU, LLVM linkers and XCode linkers as well.

Maybe we should have separate build recipes for Linux and MacOS, Linux recipe being the default? Because the only difference I can see here is how we export the dynamic symbols. Another way might be to do this:
```Makefile
EXPFLAG = -E
PTLUA_LDFLAGS = -L$(LUA_LIBDIR) $(LDFLAGS) -Wl,$(EXPFLAG)
```

And build like this for MacOS:
```
make EXPFLAG=--export_dynamic
```

But either way, we need to document a build mechanism for MacOS. I need some thought on this.

bjornbm replied:
@singul4ri7y wondering if you had time to think?

This `EXPFLAG` suggestion looks good to me. I like that it avoids the need for `-Wl,…` which is a bit cryptic for the uninitiated who just want to build with as little fuss as possible (being considerate that pallene-tracer is already introducing “fuss” into playing with pallene).

However, your code suggestion seems to be in conflict with the earlier statement:

> We should pass Linker flags to make with LDFLAGS, not via PTLUA_LDFLAGS directly

(I repeat my disclaimer: I'm not at all a Makefile or C toolchain expert, not familiar with associated conventions/best practices.)

bjornbm replied:
I added a commit with `EXPFLAG`.

singul4ri7y reviewed (changes requested):
Sorry, I completely missed the alert!

Hopefully I am up to gear on this. I have some thoughts which I commented via review.

singul4ri7y commented on Makefile:
```diff
@@ -27,7 +27,11 @@ LIBFLAG  = -fPIC -shared
+# Note: the xcode (macos) linker uses -export-dynamic instead of -E.
+# To build on macos, use make EXPFLAG=-export-dynamic
+EXPFLAG = -E
+LDFLAGS = -Wl,$(EXPFLAG)
```
I think this should be under `PTLUA_LDFLAGS` instead. Because `LDFLAGS` is also used to compile the examples, where `-Wl,-export_dynamic` is irrelevant.

singul4ri7y commented on Makefile:
```diff
@@ -65,7 +69,7 @@ clean:
 
 %.so: %.c
-	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) $(LIBFLAG) $< -o $@
+	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) $(PTLUA_LDFLAGS) $(LIBFLAG) $< -o $@ $(PTLUA_LDLIBS)
```
Do not use `$(PTLUA_LDFLAGS)` here. This flag is explicitly to build `pt-lua` only. Just using `LDFLAGS` here would suffice.

singul4ri7y reviewed (changes requested):
Looks good to me overall. Just some minor nits.

singul4ri7y commented:
@bjornbm Seems like the CI is also failing for some reason. Hopefully addressing the changes I suggested fixes it.

singul4ri7y commented:
Everything looks good. But the CI is failing for some reason. @hugomg and @bjornbm can you please take a look into it? I got midterm starting tomorrow, won't be available for couple of days.

bjornbm commented:
The problem is with the addition of `$(PTLUA_LDLIBS)` in the `.so` rule. I seems to need `-llua` on macos (or explicitly pass the `liblua.a` path) for the linking to work. But on the other hand `-llua` breaks the build on Alpine linux.

Maybe something is off with my macos install. I'll try to investigate and make sure everything is going (and looking) in the right places.

bjornbm commented:
Turns out with `-llua` earlier in the command line it is OK on both macos and Alpine linux, so I added the `-llua` in `SO_LDFLAGS` which seems appropriate. I admit I still don't know why it is required on macos but apparently not on linux. 🤷 

I also added a line to clear xcode debug outputs. They are generated due to the `-g` in `CFLAGS`.

I propose that you compare to `main` branch now and let you know if you want me to submit a cleaned up pull request.

And good luck with the midterms @singul4ri7y !

hugomg commented:
Do you know if -llua pulls in our patched version of Lua, or unpatched system Lua?

bjornbm commented:
Which liblua it finds is controlled by `LUA_PREFIX` (as shown in the README) on macos.

As mentioned, on linux it doesn't seem to matter at all. I guess it doesn't look for the symbols when creating the `.so`s? I don't know what is the difference.
